src/dbnode/x/slab/types.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package slab
+
+import (
+	"time"
+)
+
+// BufferLease is a lease for a pre-allocated buffer backed by a
+// fixed buffer manager.
+type BufferLease interface {
+	// Finalize releases the bytes returned with this lease.
+	Finalize()
+}
+
+// FixedBufferManager yields pre-alloced buffers of fixed size.
+type FixedBufferManager interface {
+	// Copy copies the source bytes into a fixed buffer, returning a copy backed
+	// by this buffer manager, and a finalizer for these bytes.
+	Copy(src []byte) ([]byte, BufferLease, error)
+	// MultiCopy copies each of the source byte slices into fixed buffers,
+	// returning copies with each slice backed by this buffer manager,
+	// and a finalizer for these bytes. These are allocated as a single block,
+	// so it should be used when multiple buffers must be allocated as a single
+	// operation to avoid deadlocking concerns.
+	MultiCopy(buffer [][]byte, sources ...[]byte) ([][]byte, BufferLease, error)
+}
+
+// Options are options for cross block iteration.
+type Options interface {
+	// Validate validates the options.
+	Validate() error
+
+	// SetFixedBufferCount sets the fixed buffer count.
+	SetFixedBufferCount(value int) Options
+
+	// FixedBufferCount returns the fixed buffer count.
+	FixedBufferCount() int
+
+	// SetFixedBufferCapacity sets the fixed buffer capacity.
+	SetFixedBufferCapacity(value int) Options
+
+	// FixedBufferCapacity returns the fixed buffer capacity.
+	FixedBufferCapacity() int
+
+	// SetFixedBufferTimeout sets the fixed buffer timeout, 0 implies no timeout.
+	SetFixedBufferTimeout(value time.Duration) Options
+
+	// FixedBufferTimeout returns the fixed buffer timeout, 0 implies no timeout.
+	FixedBufferTimeout() time.Duration
+
+	// Block()
+}
```

Comment:
can we please not use the legacy options interface? see https://github.com/uber-go/guide/blob/master/style.md#functional-options

Reply:
Honestly the proposed options there seem much clunkier to use than this style; I agree that it's better than requiring long arg lists as in the bad column there, but not certain that the proposed pattern is any easier or more flexible than this approach, especially for building options that are not expected to change much

Reply:
There are a lot of reasons that we should be preferring the other options, but one major one is that we should not be building stateful options. Options should have no opinion on ordering, use each other, do "work", etc.

Using this "options interface type" approach means that now we need to update mocks if they exist, understand if there's some special behavior in one place that's different than a similar version, there's a bunch of extra type definition, etc. I'd flip the argument on its head: I don't see any substantive value in maintaining the current options pattern versus switching to a more conventional one.

Obviously it's something that pervades a lot of the codebase, but it's something that we should be moving away from.